File: impl/ts/src/runtime/federation.ts
/**
 * Federation post-handshake dispatch loop. Mirrors
 * impl/go/internal/runtime/federation.go.
 *
 * @module
 */

import { runDispatcher, type Session } from "@sempdev/semp/session";
import type { Logger } from "pino";

import { handleFederationSubmission } from "./envelope.js";
import { handleFederationKeys } from "./keys.js";
import { handleRekeyFrame } from "./rekey.js";
import type { SQLiteKeyStore } from "../store/keys.js";
import type { SQLiteInbox } from "../store/inbox.js";
import type { SQLiteBlockList } from "../store/blocklist.js";

export interface FederationDeps {
  suite: "x25519-chacha20-poly1305" | "pq-kyber768-x25519";
  store: SQLiteKeyStore;
  inbox: SQLiteInbox;
  blockList: SQLiteBlockList;
  localDomain: string;
  domainSignFP: string;
  domainSignPriv: Uint8Array;
  domainEncFP: string;
  domainEncPriv: Uint8Array;
  domainEncPub: Uint8Array;
  identity: string;
  session: Session;
  logger: Logger | null;
}

/** Run the post-handshake message loop against an inbound peer. */
export async function serveFederation(deps: FederationDeps): Promise<void> {
  await runDispatcher(deps.session, {
    onEnvelope: async (frame) => {
      const out = await handleFederationSubmission(frame, {
        suite: deps.suite,
        store: deps.store,
        inbox: deps.inbox,
        blockList: deps.blockList,
        localDomain: deps.localDomain,
        domainSignFP: deps.domainSignFP,
        domainSignPriv: deps.domainSignPriv,
        domainEncFP: deps.domainEncFP,
        domainEncPriv: deps.domainEncPriv,
        domainEncPub: deps.domainEncPub,
        identity: deps.identity,
        session: deps.session,
        logger: deps.logger,
      });
      await deps.session.send(out);
    },
    onKeys: async (frame) => {
      const out = handleFederationKeys(frame, {
        store: deps.store,
        localDomain: deps.localDomain,
        domainSignFP: deps.domainSignFP,
        domainSignPriv: deps.domainSignPriv,
      });
      await deps.session.send(out);
    },
    onRekey: async (frame) => {
      const out = await handleRekeyFrame(
        deps.session,
        frame,
        deps.identity,
        deps.logger,
      );
      await deps.session.send(out);
    },
    onHandlerError: (err, type) => {
      deps.logger?.warn(
        { peer: deps.identity, type, err: err.message },
        "federation dispatch handler error",
      );
    },
    onFatal: (err) => {
      deps.logger?.warn(
        { peer: deps.identity, err: err.message },
        "federation dispatch fatal error",
      );
    },
  });
}

File: impl/go/internal/runtime/federation.go
package runtime

import (
	"context"
	"log/slog"

	"semp.dev/semp-go/crypto"
	"semp.dev/semp-go/delivery"
	"semp.dev/semp-go/keys"
	"semp.dev/semp-go/session"
	"semp.dev/semp-go/transport"
)

// FederationDeps groups the dependencies a federation-side dispatch
// loop needs. A federation peer forwards envelopes whose seal.signature
// was produced by the original sender domain (not by us); we verify
// against keys looked up in Store and never re-sign.
type FederationDeps struct {
	Suite     crypto.Suite
	Store     keys.Store
	Inbox     *delivery.Inbox
	BlockList delivery.BlockListLookup

	LocalDomain string

	DomainSignFP   keys.Fingerprint
	DomainSignPriv []byte
	DomainEncFP    keys.Fingerprint
	DomainEncPriv  []byte
	DomainEncPub   []byte

	// Identity is the authenticated peer identity established by the
	// preceding federation handshake (the peer server's domain).
	Identity string

	// Session is the live *session.Session backing this federation
	// connection.
	Session *session.Session

	Logger *slog.Logger
}

// ServeFederation runs the post-handshake message loop against an
// inbound federation peer. The peer is expected to forward envelopes
// whose seal.signature carries the original sender domain's
// provenance proof. This loop verifies both proofs (signature against
// the original sender domain's key from Store, session_mac against
// THIS federation session's K_env_mac), unwraps the brief, applies
// the user-level block list, and stores envelopes for local
// recipients.
//
// Federation does not multi-hop: recipients outside LocalDomain are
// reported as recipient_not_found in the per-recipient response.
func ServeFederation(ctx context.Context, conn transport.Conn, deps FederationDeps) error {
	handlers := session.DispatchHandlers{
		OnEnvelope: func(ctx context.Context, raw []byte) error {
			return handleFederationSubmissionFrame(ctx, conn, raw, deps)
		},
		OnKeys: func(ctx context.Context, raw []byte) error {
			return handleFederationKeys(ctx, conn, raw, deps)
		},
		OnRekey: func(ctx context.Context, raw []byte) error {
			return handleRekey(ctx, conn, raw, deps.Suite, deps.Session, deps.Identity, deps.Logger)
		},
		OnHandlerError: func(err error, msgType string) {
			if deps.Logger != nil {
				deps.Logger.Warn("federation dispatch handler error",
					"peer", deps.Identity,
					"type", msgType,
					"err", err,
				)
			}
		},
	}
	return session.Dispatch(ctx, conn, handlers)
}
